localization/ndt_scan_matcher/src/tree_structured_parzen_estimator.cpp

```diff
@@ -0,0 +1,233 @@
+  base_stddev_({x_stddev_, y_stddev_, z_stddev_, roll_stddev_, pitch_stddev_, yaw_stddev_})
+{
+}
+
+void TreeStructuredParzenEstimator::add_trial(const Trial & trial)
+{
+  trials_.push_back(trial);
+  std::sort(trials_.begin(), trials_.end(), [](const Trial & lhs, const Trial & rhs) {
+    return lhs.score > rhs.score;
+  });
+  good_num_ = std::count_if(trials_.begin(), trials_.end(), [this](const Trial & trial) {
+    return trial.score > MIN_GOOD_SCORE;
+  });
+  good_num_ = std::min(good_num_, static_cast<int64_t>(trials_.size() * MAX_GOOD_RATE));
+}
+
+TreeStructuredParzenEstimator::Input TreeStructuredParzenEstimator::get_next_input()
+{
+  if (trials_.empty()) {
+    // The first attempt returns all zeros.
+    // This means that the input `base_pose` is used as is.
+    // Do this because the input `base_pose` may be a good candidate.
+    return Input{};
+  } else if (static_cast<int64_t>(trials_.size()) < n_startup_trials_) {
+    // Random sampling based on prior until the number of trials reaches `n_startup_trials_`.
+    return sample_from_prior();
+  }
+
+  std::vector<double> scores;
+  for (int64_t i = 0; i < good_num_; i++) {
+    const double diff = trials_[i].score - MIN_GOOD_SCORE;
+    scores.push_back(std::exp(diff));
+  }
+  constexpr double PRIOR_WEIGHT = 1.0;
+  scores.push_back(PRIOR_WEIGHT);
+  std::discrete_distribution<int64_t> dist_good(scores.begin(), scores.end());
+
+  Input best_input;
+  double best_score = -1e9;
+  for (int64_t i = 0; i < N_EI_CANDIDATES; i++) {
+    Input input{};
+    const int64_t index = dist_good(engine_);
+    if (index == good_num_) {
+      input = sample_from_prior();
+    } else {
+      const Input & base = trials_[index].input;
+
+      // When score is TARGET_SCORE, coeff is 1.0 / 40
+      // When score is MIN_GOOD_SCORE, coeff is 5.0 / 40
+      // Interpolate linearly between these two values.
+      const double score_diff = std::max(TARGET_SCORE - trials_[index].score, 0.0);
+      const double linear = 1.0 + (score_diff / (TARGET_SCORE - MIN_GOOD_SCORE)) * 4.0;
+      const double coeff = BASE_STDDEV_COEFF * linear / 40;
+      input.x = base.x + dist_normal_(engine_) * coeff * x_stddev_;
+      input.y = base.y + dist_normal_(engine_) * coeff * y_stddev_;
+      input.z = base.z + dist_normal_(engine_) * coeff * z_stddev_;
+      input.roll = base.roll + dist_normal_(engine_) * coeff * roll_stddev_;
+      input.pitch = base.pitch + dist_normal_(engine_) * coeff * pitch_stddev_;
+      input.yaw = base.yaw + dist_normal_(engine_) * coeff * yaw_stddev_;
+
+      // fixed angle
+      input.roll = fix_angle(input.roll);
+      input.pitch = fix_angle(input.pitch);
+      input.yaw = fix_angle(input.yaw);
+    }
+    const double score = acquisition_function(input);
+    if (score > best_score) {
+      best_score = score;
+      best_input = input;
+    }
+  }
+  return best_input;
+}
+
+double TreeStructuredParzenEstimator::acquisition_function(const Input & input)
```

Comment:
Could you elaborate more on this function?
Maybe renaming the function improves a lot, since `acquisition_function` sounds very general. (Not sure if I understood correctly, but something like `get_predicted_score_from_prior_results` for instance.)

Reply:
The term `acquisition_function` is often used in the context of Bayesian optimization. To be precise, what is implemented here is a function that returns the score of the selection criteria resulting from the transformation of Expectation Improvement (EI), a kind of acquisition functions. So naming it is a difficult task, but I think it is not so bad.

Reply:
But I renamed to make it better.
https://github.com/autowarefoundation/autoware.universe/pull/5078/commits/8928c88c1df03776dc496b358d15c1fbfb1ee4d6